spatial_batch_kernels.hpp

```diff
@@ -0,0 +1,1122 @@
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#ifndef VLASIATOR_SPATIAL_BATCH_KERNELS_HPP
+#define VLASIATOR_SPATIAL_BATCH_KERNELS_HPP
+
+#ifdef USE_WARPACCESSORS
+ #define USE_BATCH_WARPACCESSORS
+#endif
+/** GPU kernel for identifying which blocks have relevant content */
+__global__ void batch_update_velocity_block_content_lists_kernel (
+   vmesh::VelocityMesh **vmeshes,
+   vmesh::VelocityBlockContainer **blockContainers,
+   Hashinator::Hashmap<vmesh::GlobalID,vmesh::LocalID>** allMaps,
+   Real* velocity_block_min_values,
+   const bool gatherMass,
+   Real* dev_mass
+   ) {
+   // launch griddim3 grid(launchBlocks,nCells,1);
+   const uint nCells = gridDim.y;
+   const int cellIndex = blockIdx.y;
+   const int blocki = blockIdx.x;
+   const uint ti = threadIdx.x;
+
+   vmesh::VelocityMesh* vmesh = vmeshes[cellIndex];
+   vmesh::VelocityBlockContainer* blockContainer = blockContainers[cellIndex];
+   Real velocity_block_min_value = velocity_block_min_values[cellIndex];
+   Hashinator::Hashmap<vmesh::GlobalID,vmesh::LocalID>* vbwcl_map = allMaps[cellIndex];
+   Hashinator::Hashmap<vmesh::GlobalID,vmesh::LocalID>* vbwncl_map = allMaps[nCells+cellIndex];
+
+   // Each GPU block / workunit can theoretically manage several Vlasiator velocity blocks at once.
+   const uint vlasiBlocksPerWorkUnit = 1;
+   const uint workUnitIndex = 0; // [0,vlasiBlocksPerWorkUnit)
+   // const uint vlasiBlocksPerWorkUnit = WARPSPERBLOCK * GPUTHREADS / WID3;
+   // const uint workUnitIndex = ti / WID3; // [0,vlasiBlocksPerWorkUnit)
+   const uint b_tid = ti % WID3; // [0,WID3)
+   const uint blockLID = blocki * vlasiBlocksPerWorkUnit + workUnitIndex; // [0,nBlocksToChange)
+
+   __shared__ int has_content[WARPSPERBLOCK * GPUTHREADS];
+   __shared__ Real gathered_mass[WARPSPERBLOCK * GPUTHREADS];
+   const uint nBlocks = vmesh->size();
+#ifdef DEBUG_SPATIAL_CELL
+   if (nBlocks != blockContainer->size()) {
+      if (b_tid==0) printf("VBC and vmesh size mismatch in batch_update_velocity_block_content_lists_kernel!\n");
+      return;
+   }
+#endif
+   if (blockLID < nBlocks) {
+      const vmesh::GlobalID blockGID = vmesh->getGlobalID(blockLID);
+#ifdef DEBUG_SPATIAL_CELL
+      if (blockGID == vmesh->invalidGlobalID()) {
+         if (b_tid==0) printf("Invalid GID encountered in batch_update_velocity_block_content_lists_kernel!\n");
+         return;
+      }
+      if (blockLID == vmesh->invalidLocalID()) {
+         if (b_tid==0) printf("Invalid LID encountered in batch_update_velocity_block_content_lists_kernel!\n");
+         return;
```

Comment:
Shouldn't these guys crash? Or can't they in a kernel?

Reply:
In-kernel one could indeed assert.